Mission2 함수형 todolist

YeonjuOHYE opened the pull request:
## 내용
#### todos.js 구현
#### (선택 미션)
update 출력 2초 delay

## 참고 사항 
readline 쪽 사용방법이 미숙하여서 update를 입력하면 delay로 인해
```bashshell
명령어 입력하시오 update$$1$$todo
명령어 입력하시오 Map {
  1 => { type: 'todo', name: '두번째 할 일', tags: [] },
  2 => { type: 'doing', name: '세번째 할 일', tags: [] } }
[todo]:1,[doing]:2
```
명령어 입력하시오 가 등장하지 않습니다...(그래도 입력은 됩니다.)

Commit:
skeleton.js

Commit:
//todo.js

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+todoFuncs.set("add", add);
+
+const update = (id, type) =>
+  pipe(
```
pipe안에 인자로 들어가는 
함수들의 선언을 외부에 해보세요.

pipe(setHistory, delay)(); 

그러면 이런식으로 호출될 수도 있을거 같아요.

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+          resolve();
+        }, 2000);
+      }),
+    deplyPromise => deplyPromise.then(showCurrent)
```
delay이후에 readline 입력을 계속 받으려면,,
처음부터 async/await 함수를 처음호출되는 부분부터 쭉 ~ 사용하거나, 
아니면 readline 객체를 전달하는 방법도 있을거 같습니다.

추가로 무언가 하시려면 이부분을 꼭 해보시는 것을 권장합니다 ^^
비동기개념이 섞여 있어서 중요합니다.

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+  )();
+
+const showType = type =>
+  pipe(
```
복잡한 로직을 추상화시켜서 감춰놓는 것이 함수형프로그래밍의 장점인거 같아요.

이미 잘 구현하셨지만, 
pipe안에 함수로직을 모두 표현하지 않으시고 분리해서 선언해두시면 될 거 같습니다.

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+      reulstStr = `총 ${typeDataList.length}: `;
+      return (reulstStr += typeDataList
+        .map(({ name, id }) => {
+          return "'" + name + "," + id + "번'";
```
문자열 조작은 이제 template literal을 사용해보세요!

crongro replied:
reduce는 배열을 객체로, 배열을 문자열(hTML문자열등)으로 바꾸는데 유용합니다.
새로운 차원의 타입으로 변경할때 매우 유용해요 
이부분에 한번 적용을 해보시죠! (map.join 대신에 reduce로!)

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+  DONE: "done"
+};
+
+let firstHistory = new Map().set(0, {
```
1~n번째 할일을 중복으로 생성하셨는데요.
생성자패턴을 활용하는 방법을 해보시면 더 좋겠어요.
```javascript
function Todo() {
}
const todo = new Todo({....});
```

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+  pipe(
+    getCurrentData,
+    currentData => {
+      lastKey = Array.from(currentData.keys()).reduce((prev, cur) =>
```
Array.from도 명시적으로 표현해서 좋죠.
spread operator 를 사용해도 되긴합니다.

crongro commented on todo.js:
```diff
@@ -0,0 +1,205 @@
+
+const groupByType = data => {
+  console.log(data);
+  return Array.from(data.keys()).reduce((accumulator, val) => {
```
아 이부분에서 reduce를 잘 사용하셨네요
groupByType 코드를 보니 immutable도 잘 이해하신 거 같고요.
잘 구현하신 코드 같습니다.

crongro reviewed (changes requested):
전체적으로 ES6 문법을 잘 활용한 부분이 좋습니다.
함수들의 크기가 좋지만, pipe안에서는 함수이름만 표현되면 더 좋겠고요.

비동기 부분은 음 꼭 이해하고 해결하시면 좋겠어요 ^^
일단 PR을 request changes 로 남겨둘게요!

Commit:
리뷰 반영
- delay이후에 readline 입력을 계속 받도록
- Todo 생성자 사용
- .join() 사용하지 않고 reduce 사용하도록
- tmplate literal 사용하도록

Commit:
#mission3

Commit:
read-me update
